### PR TITLE
Add a build variant to link against Pocl on OSX

### PR DESCRIPTION
**Pull request by inducer**

### CODE REVIEW

**Comment by inducer** [comment]
cc @mattwala

**Comment by inducer** [comment]
@mattwala Any clue what's going on here? Why is it referring to `_clCompileProgram`, not `clCompileProgram`?

```
ImportError: dlopen(/Users/travis/miniconda3/conda-bld/pyopencl_1487400667425/_t_env/lib/python3.6/site-packages/pyopencl/_cffi.cpython-36m-darwin.so, 2): Symbol not found: _clCompileProgram

  Referenced from: /Users/travis/miniconda3/conda-bld/pyopencl_1487400667425/_t_env/lib/python3.6/site-packages/pyopencl/_cffi.cpython-36m-darwin.so

  Expected in: flat namespace

 in /Users/travis/miniconda3/conda-bld/pyopencl_1487400667425/_t_env/lib/python3.6/site-packages/pyopencl/_cffi.cpython-36m-darwin.so
```

**Comment by mattwala** [comment]
Pretty sure the `_` is just a mangling prefix.

Bizarrely, that symbol isn't exported in libpocl:
```
(py3) $ nm -a /Users/matt/miniconda3/envs/py3/lib/libpocl.1.dylib  | grep clCompileProgram
(py3) $ 
```

**Comment by mattwala** [comment]
Actually, `clCompileProgram` doesn't seem to exist in pocl at all.

**Comment by mattwala** [comment]
This seems to do it: http://stackoverflow.com/a/27728404. In your own words :)

**Comment by inducer** [comment]
Thanks! (Also, ouch. :) One remaining (mini-)challenge is that pocl supports a curious mixture of not-quite CL 1.1 and CL 2.0 (especially SVM). Ideally, we wouldn't lose the latter...

**Comment by inducer** [comment]
@mattwala 
```
test_algorithm.py ..................................xxx../Users/travis/miniconda3/conda-bld/pyopencl_1487481660740/test_tmp/run_test.sh: line 3:  3362 Bus error: 10           python -m pytest
```
Any clues?

**Comment by mattwala** [comment]
> Any clues?

I suspect that's the crash in bitonic sort that I was seeing. I still don't know what caused it.

File: recipe/meta.yaml [hunk 1]
```diff
@@ -1,6 +1,7 @@
 {% set name = "pyopencl" %}
 {% set version = "2016.2.1" %}
 {% set sha256 = "3fcb59ab9c85e08d96a24388a736cc3d0bbd9608efff96ecb25d3124fde6f4b7" %}
+{% set osx_variant = os.environ.get('OSX_VARIANT', 'default') %}
```
Comment (jakirkham) [inline recipe/meta.yaml]: Should this have been `"native"`?

File: recipe/meta.yaml [hunk 2]
```diff
@@ -23,6 +39,11 @@ requirements:
         - ocl-icd  # [linux]
         - khronos-opencl-icd-loader  # [win]
         - mingwpy  # [win]
+{% if osx_variant == "pocl" %}
+        - osx-pocl-opencl
+        - pocl
```
Comment (jakirkham) [inline recipe/meta.yaml]: Asking out of ignorance. Would it make sense to have `pocl` as an option on Linux?

File: recipe/meta.yaml (hunk 1 above)
Comment (inducer) [inline recipe/meta.yaml]: https://github.com/inducer/pyopencl-feedstock/commit/6734bfac72f7cb0a16d14509d3ace87d95ed188f

Thx!

File: recipe/meta.yaml (hunk 2 above)
Comment (inducer) [inline recipe/meta.yaml]: No, not really. Under normal circumstances (i.e. non-Apple), it is the ICD loader's (e.g. `ocl-icd`) responsibility to dispatch to the right implementation. Anything using CL only needs to link against the loader and doesn't need to entangle itself with actual implementation choice. Instead, this choice can safely be deferred to run time. I.e. on Linux, simply install `pocl` and `ocl-icd`, and they'll automatically see and talk to each other. Plus you could install more CL implementations (or add `.icd` files--simple text files with file names--for your built-in GPUs to `$CONDA/etc/OpenCL/vendors`), and they would automatically work.

Apple (AFAICT) doesn't ship a usable ICD loader, and ocl-icd won't compile on Apple, so there's nothing to fill the "runtime implementation multiplexer" role. That's why there's this link-directly-to-pocl crap in here. It's expedient, but it goes against CL's (fairly elegant) multiplexed design.

Comment (jakirkham) [inline recipe/meta.yaml]: Ah! That makes more sense. Clever design.

Leave it to Apple to buck the curve. Though I'm sure they have their own reasons for it.

**Pull request merged**